lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -1062,22 +1075,96 @@ def get_service_status(  # noqa: C901
+        if len(contents) > 1 and contents[0].credentials.s3 != contents[1].credentials.s3:
+            return None
+        return contents[0].credentials.s3
+
+    @override
+    def is_relation_set(self) -> bool:
+        """Checks if the relation is set for the plugin handler."""
+        if isinstance(self.backup(), OpenSearchBackup):
+            relation = self._charm.model.get_relation(self._relation_name)
+            return relation is not None and relation.units != {}
+
+        # We are not not the MAIN_ORCHESTRATOR
+        # The peer-cluster relation will always exist, so we must check if
+        # the relation has the information we need or not.
+        return (
+            self._get_peer_rel_data is not None
+            and self._get_peer_rel_data.access_key
+            and self._get_peer_rel_data.secret_key
+        )
+
+    @property
+    def _relation_name(self) -> str:
+        rel_name = PeerClusterRelationName
+        if isinstance(self.backup(), OpenSearchBackup):
+            rel_name = S3_RELATION
+        return rel_name
+
+    @override
+    def get_relation_data(self) -> Dict[str, Any]:
+        """Returns the relation that the plugin manager should listen to."""
+        if not self.is_relation_set():
+            return {}
+        elif isinstance(self.backup(), OpenSearchBackup):
+            relation = self._charm.model.get_relation(self._relation_name)
+            return relation and relation.data.get(relation.app, {})
+        return self._get_peer_rel_data.dict()
+
+    def backup(self) -> OpenSearchBackupBase:
+        """Implements the logic that returns the correct class according to the cluster type."""
+        if not OpenSearchBackupFactory._backup_obj:
+            OpenSearchBackupFactory._backup_obj = self._get_backup_obj()
+        return OpenSearchBackupFactory._backup_obj
+
+    def _get_backup_obj(self) -> OpenSearchBackupBase:
+        """Implements the logic that returns the correct class according to the cluster type.
+
+        This class is solely responsible for the creation of the correct S3 client manager.
+
+        If this cluster is an orchestrator or failover cluster, then return the OpenSearchBackup.
+        Otherwise, return the OpenSearchNonOrchestratorBackup.
+
+        There is also the condition where the deployment description does not exist yet. In this
+        case, return the base class OpenSearchBackupBase. This class solely defers all s3-related
+        events until the deployment description is available and the actual S3 object is allocated.
+        """
+        if not self._charm.opensearch_peer_cm.deployment_desc():
+            # Temporary condition: we are waiting for CM to show up and define which type
+            # of cluster are we. Once we have that defined, then we will process.
+            return OpenSearchBackupBase(self._charm)
+        elif (
+            self._charm.opensearch_peer_cm.deployment_desc().typ
```

Comment:
this will need to change to `self._charm.opensearch_peer_cm.is_provider(typ="main")`

Reply:
@Mehdi-Bendriss is this part of the new merged PR?

Reply:
As discussed, in case we are interested in the large deployments only. The new API set: `self._charm.opensearch_peer_cm.is_{provider|consumer}` is more suitable.

If we *do not* want to differ between them, then we use the `self._charm.opensearch_peer_cm.deployment_desc().typ`.

The advantage of the provider / consumer: checks if the relation is stablished, has data, ready for consumption, etc.